File: old/mailtidy/integrations/email/mock.py
"""本地开发 / 单测用的假邮箱实现。

特点：
- 内置一组覆盖常见分类的样例邮件（重要 / newsletter / 促销 / 通知 / 账单）。
- 写动作不真的修改邮件，只把操作字符串追加到 ``self.operations`` 里，
  让单测可以断言"是否调用过 archive:m3"。
"""

from __future__ import annotations

from datetime import datetime, timedelta

from mailtidy.data.models import EmailMessage
from mailtidy.integrations.email.base import EmailConnector


class MockEmailConnector(EmailConnector):
    """固定 6 封样例邮件、把所有写动作记录到字符串列表里的本地 connector。"""

    def __init__(self) -> None:
        now = datetime.now()
        # 6 封样例邮件，故意覆盖每种分类，让 run-cleanup demo 一次跑全所有分支
        self.messages = [
            EmailMessage("m1", "ceo@example.com", "Need your approval today", "Please approve the Q2 budget by 5pm.", now - timedelta(hours=1)),
            EmailMessage("m2", "news@productweekly.com", "Product Weekly #182", "AI tools, SaaS metrics, and three long reads.", now - timedelta(hours=2)),
            EmailMessage("m3", "deals@shop.example", "70% off today only", "Flash sale ends tonight.", now - timedelta(hours=3)),
            EmailMessage("m4", "github@github.com", "[MailTidy] CI failed", "Build failed on main branch.", now - timedelta(hours=4)),
            EmailMessage("m5", "billing@notion.so", "Your Notion receipt", "Payment receipt: you were charged $10.00 for your monthly plan.", now - timedelta(days=10), body="Your plan: Plus. Amount: $10.00 monthly."),
            EmailMessage("m6", "billing@netflix.com", "Your Netflix payment", "Payment receipt: your card was charged $15.99.", now - timedelta(days=20), body="Your plan: Premium. Amount: $15.99 monthly."),
        ]
        self.operations: list[str] = []

    def fetch_recent(self, hours: int = 24, limit: int = 200, unread_only: bool = True) -> list[EmailMessage]:
        cutoff = datetime.now() - timedelta(hours=hours)
        messages = [message for message in self.messages if message.date >= cutoff]
        if unread_only:
            messages = [message for message in messages if message.unread]
        return messages[:limit]

    def search(self, query: str, months: int = 6) -> list[EmailMessage]:
        # 去掉外层引号，让 ``"payment receipt"`` 这样的查询能正确匹配
        phrase = query.strip().strip('"').lower()
        cutoff = datetime.now() - timedelta(days=months * 30)
        results = []
        for message in self.messages:
            haystack = f"{message.sender} {message.subject} {message.snippet} {message.body}".lower()
            if message.date >= cutoff and phrase in haystack:
                results.append(message)
        return results

    def archive(self, email_ids: list[str]) -> None:
        self.operations.append(f"archive:{','.join(email_ids)}")

    def label(self, email_ids: list[str], label: str) -> None:
        self.operations.append(f"label:{label}:{','.join(email_ids)}")

    def star(self, email_ids: list[str]) -> None:
        self.operations.append(f"star:{','.join(email_ids)}")

    def mark_read(self, email_ids: list[str]) -> None:
        self.operations.append(f"mark_read:{','.join(email_ids)}")

    def save_draft(self, email_id: str, body: str) -> None:
        # 截断到 24 字符，避免单测断言时受 LLM 输出长短影响
        self.operations.append(f"draft:{email_id}:{body[:24]}")


__all__ = ["MockEmailConnector"]

File: src/integrations/email/mock.ts
/**
 * 本地开发 / 单测用的假邮箱实现。
 *
 * - 内置一组覆盖常见分类的样例邮件（重要 / newsletter / 促销 / 通知 / 账单）。
 * - 写动作不真的修改邮件，只把操作字符串追加到 `operations` 里，让单测可以
 *   断言"是否调用过 archive:m3"。
 */

import type { EmailConnector, FetchRecentOptions } from "./base.js";
import type { EmailMessage } from "../../data/models.js";

const HOUR_MS = 60 * 60 * 1000;
const DAY_MS = 24 * HOUR_MS;

function isoAgo(ms: number): string {
  return new Date(Date.now() - ms).toISOString();
}

export class MockEmailConnector implements EmailConnector {
  readonly operations: string[] = [];
  readonly messages: EmailMessage[] = [
    {
      id: "m1",
      sender: "ceo@example.com",
      subject: "Need your approval today",
      snippet: "Please approve the Q2 budget by 5pm.",
      date: isoAgo(1 * HOUR_MS),
      unread: true,
    },
    {
      id: "m2",
      sender: "news@productweekly.com",
      subject: "Product Weekly #182",
      snippet: "AI tools, SaaS metrics, and three long reads.",
      date: isoAgo(2 * HOUR_MS),
      unread: true,
    },
    {
      id: "m3",
      sender: "deals@shop.example",
      subject: "70% off today only",
      snippet: "Flash sale ends tonight.",
      date: isoAgo(3 * HOUR_MS),
      unread: true,
    },
    {
      id: "m4",
      sender: "github@github.com",
      subject: "[MailTidy] CI failed",
      snippet: "Build failed on main branch.",
      date: isoAgo(4 * HOUR_MS),
      unread: true,
    },
    {
      id: "m5",
      sender: "billing@notion.so",
      subject: "Your Notion receipt",
      snippet: "Payment receipt: you were charged $10.00 for your monthly plan.",
      date: isoAgo(10 * DAY_MS),
      unread: true,
      body: "Your plan: Plus. Amount: $10.00 monthly.",
    },
    {
      id: "m6",
      sender: "billing@netflix.com",
      subject: "Your Netflix payment",
      snippet: "Payment receipt: your card was charged $15.99.",
      date: isoAgo(20 * DAY_MS),
      unread: true,
      body: "Your plan: Premium. Amount: $15.99 monthly.",
    },
  ];

  async fetchRecent(options: FetchRecentOptions = {}): Promise<EmailMessage[]> {
    const { hours = 24, limit = 200, unreadOnly = true } = options;
    const cutoff = Date.now() - hours * HOUR_MS;
    let items = this.messages.filter((m) => new Date(m.date).getTime() >= cutoff);
    if (unreadOnly) items = items.filter((m) => m.unread);
    return items.slice(0, limit);
  }

  async search(query: string, months: number = 6): Promise<EmailMessage[]> {
    // 去掉外层引号，让 `"payment receipt"` 这样的查询能正确匹配。
    const phrase = query.trim().replace(/^"|"$/g, "").toLowerCase();
    const cutoff = Date.now() - months * 30 * DAY_MS;
    return this.messages.filter((m) => {
      const haystack = `${m.sender} ${m.subject} ${m.snippet} ${m.body ?? ""}`.toLowerCase();
      return new Date(m.date).getTime() >= cutoff && haystack.includes(phrase);
    });
  }

  async readById(emailId: string): Promise<EmailMessage | null> {
    return this.messages.find((message) => message.id === emailId) ?? null;
  }

  async archive(emailIds: string[]): Promise<void> {
    this.operations.push(`archive:${emailIds.join(",")}`);
  }
  async label(emailIds: string[], label: string): Promise<void> {
    this.operations.push(`label:${label}:${emailIds.join(",")}`);
  }
  async star(emailIds: string[]): Promise<void> {
    this.operations.push(`star:${emailIds.join(",")}`);
  }
  async markRead(emailIds: string[]): Promise<void> {
    this.operations.push(`mark_read:${emailIds.join(",")}`);
  }
  async saveDraft(emailId: string, body: string): Promise<void> {
    // 截断到 24 字符，避免单测断言时受 LLM 输出长短影响
    this.operations.push(`draft:${emailId}:${body.slice(0, 24)}`);
  }
}
